tests/src/test/scala/BinaryTreeTest.scala

```diff
@@ -0,0 +1,235 @@
+
+    emptyTree should be(BTNil())
+  }
+
+  it  should "return a branch with one item when add to empty tree" in {
+
+    val empty = BinaryTree.empty[Int]
+
+    val branch = empty.add(5)
+
+    branch.value should be(Some(5))
+  }
+
+  it should "insert value into left branch when add to brach to the left" in {
+
+    val branch = BinaryTree(5)
+
+    val result = branch.add(3)
+
+    result.value should be(Some(5))
+    result.left.value should be(Some(3))
+  }
+
+  it should "insert value into right branch when add to branch to the right" in {
+
+    val branch = BinaryTree(5)
+
+    val result = branch.add(7)
+
+    result.value should be(Some(5))
+    result.right.value should be(Some(7))
+  }
+
+  it should "balance the tree after item added" in {
+    val tree = BinaryTree(3) + 2 + 1
+
+    tree.value should be(Some(2))
+    tree.left.value should be(Some(1))
+    tree.right.value should be (Some(3))
+
+    val second = tree + 4
+
+    second.right.right.value should be(Some(4))
+
+    val third = second + 5
+    third.right.value should be(Some(4))
+    third.right.left.value should be(Some(3))
+    third.right.right.value should be(Some(5))
+  }
+
+
+
+}
+
+class BinaryTreeTestJoins extends FlatSpec with Matchers {
+  implicit val cmp =  new Order[Int] {
+    override def apply(l: Int, r: Int): Ordering = l.compareTo(r) match {
+      case 0  => Order.EQ
+      case -1 => Order.LT
+      case 1  => Order.GT
+    }
+  }
+
+  "Binary Tree" should "be the same after joined with an empty tree" in {
+    val tree = BinaryTree(5) + 3 + 2 + 7 + 9
+
+    val empty = BinaryTree.empty[Int]
+
+    val joined = tree ++ empty
+
+    ScalaListConverter.toScalaList(tree.toLst) should contain inOrderOnly(2, 3, 5, 7, 9)
+  }
+
+  it should "the same plus items from the other tree" in {
+    val tree = BinaryTree(5) + 3 + 2 + 7 + 9
+    val another = BinaryTree(1) + 4 + 8
+
+    val joined = tree ++ another
+
+    ScalaListConverter.toScalaList(joined.toLst) should contain inOrder(1, 2, 3, 4, 5, 7, 8, 9)
+  }
+}
+
+class BinaryTreeTestWalk extends FlatSpec with Matchers {
+  implicit val cmp =  new Order[Int] {
+    override def apply(l: Int, r: Int): Ordering = l.compareTo(r) match {
+      case 0  => Order.EQ
+      case -1 => Order.LT
+      case 1  => Order.GT
+    }
+  }
+
+   "Binary Tree"  should "return sorted seq when toLst" in {
+    val tree = BinaryTree(5).add(2).add(1).add(3)
+
+    val sorted = tree.toLst
+
+    ScalaListConverter.toScalaList(sorted) should contain inOrder(1, 2, 3, 5)
+  }
+
+  it  should "return sorted seq when walk in-order" in {
+    val tree = BinaryTree(5).add(2).add(1).add(3)
+
+    val sorted = tree.inorder
+
+    sorted should be (tree.toLst)
+  }
+
+  it  should "return node-left-right for each node when walk in pre-order" in {
+    val tree = BinaryTree(5).add(2).add(1).add(3)
+
+    val preorder = tree.preorder
+
+    ScalaListConverter.toScalaList(preorder) should contain inOrder(2, 1, 5, 3)
+  }
```

Comment:
this seems very arbitrary, it seems like you crafted the test to make sure it passes, there are plenty of other ordered balanced trees that wouldn't have this ordering.

You should be crafting tests which make sure that don't know about what the implementation of the algorithm is, but instead just test that any invariants the algorithm have still hold.

so in this case we are testing that inserting elements in a certain order got some exact tree. but why do we care? all we should care about is that the tree is ordered and balanced.